### PR TITLE
Add support for several elasticsearch auth options, TLS and more options

### PR DESCRIPTION
**Pull request by mfelsche**
# Pull request

## Description

New options for `elastic` connector:

- `headers`
- `timeout`
- `auth`
- `tls`

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

### CODE REVIEW

File: src/connectors/impls/elastic.rs [hunk 1]
```diff
@@ -54,6 +82,10 @@ fn default_concurrency() -> usize {
     DEFAULT_CONCURRENCY
 }
 
+fn default_auth() -> Auth {
```
Comment (Licenser) [inline src/connectors/impls/elastic.rs:85]: could we make Auth 'Default'? that way we avoid extra functions for this

File: src/connectors/impls/http/auth.rs [hunk 2]
```diff
@@ -12,14 +12,30 @@
...
 /// Authorization methods
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
 #[serde(rename_all = "lowercase")]
 pub enum Auth {
     #[serde(alias = "basic")]
     Basic { username: String, password: String },
+    #[serde(alias = "bearer")]
+    Bearer(String),
+    #[serde(alias = "client_certificate")]
+    ClientCertificate {
```
Comment (Licenser) [inline src/connectors/impls/http/auth.rs:31]: Would make sense to move this into `utils::tls` and keep the same naming conventions we have in other TLS related configs?

Comment (mfelsche) [reply src/connectors/impls/http/auth.rs:31]: Why move it out of this enum? Or would you wrap this struct in a newtype and move the struct to tls?

Comment (Licenser) [reply src/connectors/impls/http/auth.rs:31]: Ja I was thinking to create a struct fir client certificates, since they do exist for all TLS endpoints, not just HTTP, and wrap this here as a serde-flattened struct. Moving it to the `tls` utils will make it easier ti keep naming uniform through places, like not having `private_key` in one place and `key` in another

File: src/connectors/impls/http/auth.rs [hunk 3]
```diff
@@ -12,14 +12,30 @@
...
+        /// path to a file containing one or more certificates in DER encoding that form a chain, ordered from leaf to root cert
+        certificate_chain: String,
+        /// path to a file containing the private key in either RSA or PKCS8 encoded form
+        private_key: String,
+    },
+    #[serde(alias = "elastic_api_key")]
```
[the remark this answers is not in the document]
Comment (Licenser) [reply src/connectors/impls/http/auth.rs:37]: I'm worried that this will grow to a bunch of sink specific implementations that will end up being confusing.

File: src/connectors/impls/http/auth.rs (hunk 2 above)
Comment (mfelsche) [reply src/connectors/impls/http/auth.rs:31]: Yeah makes sense. Will do.

File: src/connectors/impls/http/auth.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Licenser) [reply src/connectors/impls/http/auth.rs:37]: I mean if we decide that's what we want, 👍 , but lets not have that creep in pice by pice without a conscious decision to make it this.

File: src/connectors/impls/elastic.rs (hunk 1 above)
Comment (mfelsche) [reply src/connectors/impls/elastic.rs]: Done

File: src/connectors/impls/http/auth.rs (hunk 2 above)
Comment (mfelsche) [reply src/connectors/impls/http/auth.rs]: Done, `Auth` is no solely for populating the `Authorization` header.

File: src/connectors/impls/http/auth.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mfelsche) [reply src/connectors/impls/http/auth.rs:25]: I don't think this is too confusing. It is some special kind of auth that elastic pulled, and it actually is usable with http as well. I don't have a good idea how to refactor this one enum variant so it is easy to use and self-explanatory. I wouldn't want users to have to base64 their config on their own just for us having some leaner code. I would vote for keeping it this way.

[the remark this answers is not in the document]
Comment (Licenser) [reply src/connectors/impls/http/auth.rs:25]: works 👍

**Review by Licenser — APPROVED** [review]
🚀

File: src/postprocessor.rs [hunk 4]
```diff
@@ -61,6 +61,7 @@ pub fn lookup_with_config(config: &PostprocessorConfig) -> Result<Box<dyn Postpr
     match config.name.as_str() {
         "compress" => Ok(Box::new(Compress::from_config(config.config.as_ref())?)),
         "separate" => Ok(Box::new(separate::Separate::from_config(&config.config)?)),
+        "lines" => Ok(Box::new(separate::Separate::default())),
```
Comment (Licenser) [inline src/postprocessor.rs:64]: This kind of doesn't belong in this PR and having it twice isn't really great.

**Review by Licenser — APPROVED** [review]
🚀 🇬🇵

**Review by Licenser — APPROVED** [review]
I shall not be dismissed like this! :joy: (or at least my approval shall not be dismissed :P)

**Comment by mfelsche** [comment]
it is not working yet anyways. For some reason on CI the elasticsearch docker image is not coming up properly with http ssl settings :( Still investigating with some heavy printf debugging

**Review by Licenser — DISMISSED** [review]
Not today github! Approved!

**Review by Licenser — APPROVED** [review]
Approval!
